[WIP]: Initial yocto support

bhargavdas opened the pull request:
Yocto recipes for mtda.

**Yocto Compatibilty**
Release: v3.4 , honister (since the overriding syntax has changed and are not backward compatible)

**What's been done**
- Created new recipes for some of the python libraries and also for mtda.
- New image named mtda-image, based on core-image-minimal.
- Services are pushed to rootfs from the recipes unlike isar.

**Status**
- Work in progress.
- Though the mtda recipe's PV is 0.16 but currently the recipe is using the tip of the master branch (for development).

**What works**
- mtda-cli
- mtda service
- Other basic features

**What doesn't work**
- Switching of mass storage device from target to host (and vice versa).

**Tested on**
- imx6dlsdb

**Maintenance Strategy** (suggestion)
- Update the recipes once a tc1 release tag is assigned, and start testing.

**Thing to review now for Maintainers**
- Check the licenses in the recipes.
- Is the maintenance strategy acceptable ?
- Review the layer name if it is appropraite
- Review the recipes
- Check for unwanted bits

vj-kumar commented:
@bhargavdas Wondering if we can start with a proven mtda platform for the initial support. Beaglebone black might be a good candidate to start with. All you need is the poky repo.

Starting with such a platform will give us some idea on how much of things we can reuse, and how we can align our existing implementation to achieve that.

vj-kumar commented:
> @bhargavdas Wondering if we can start with a proven mtda platform for the initial support. Beaglebone black might be a good candidate to start with. All you need is the poky repo.
> 
> Starting with such a platform will give us some idea on how much of things we can reuse, and how we can align our existing implementation to achieve that.

https://github.com/vj-kumar/mtda/blob/vijai/yocto/kas/yocto/bbb.yml

Initial yocto bbb yml file. Builds fine.

bhargavdas commented:
> @bhargavdas Wondering if we can start with a proven mtda platform for the initial support. Beaglebone black might be a good candidate to start with. All you need is the poky repo.
> 
> Starting with such a platform will give us some idea on how much of things we can reuse, and how we can align our existing implementation to achieve that.

I agree, but I don't have a beaglebone black to test. Though I can build an image using this layer and ask someone to test it.

vj-kumar commented:
> > @bhargavdas Wondering if we can start with a proven mtda platform for the initial support. Beaglebone black might be a good candidate to start with. All you need is the poky repo.
> > Starting with such a platform will give us some idea on how much of things we can reuse, and how we can align our existing implementation to achieve that.
> 
> I agree, but I don't have a beaglebone black to test. Though I can build an image using this layer and ask someone to test it.

I have one and I can lend it to you.

bhargavdas commented:
> > > @bhargavdas Wondering if we can start with a proven mtda platform for the initial support. Beaglebone black might be a good candidate to start with. All you need is the poky repo.
> > > Starting with such a platform will give us some idea on how much of things we can reuse, and how we can align our existing implementation to achieve that.
> > 
> > 
> > I agree, but I don't have a beaglebone black to test. Though I can build an image using this layer and ask someone to test it.
> 
> I have one and I can lend it to you.

Sounds good :)

vj-kumar commented on meta-mtda/recipes-devtools/python/python3-ifaddr_0.1.7.bb:
```diff
@@ -0,0 +1,14 @@
+DESCRIPTION = "Cross-platform network interface and IP address enumeration \
```
Can we reuse https://layers.openembedded.org/layerindex/recipe/164913/ instead?

bhargavdas replied:
This is actually pulled from the link you have mentioned. This was done to keep dependencies in one place.

bhargavdas replied:
Yes, we can use the upstream recipe instead of keep in this layer.
Only minor concern is if the upstream recipe's version is bumped it should not affect the functionality in the mtda code.

vj-kumar replied:
You can make sure of that by having the revisions captured in the kas file. And bump only if needed. I don't see a need to bump very often. We should reuse recipes from other layers as possible.

vj-kumar commented on meta-mtda/recipes-devtools/python/python3-lockfile_0.12.2.bb:
```diff
@@ -0,0 +1,17 @@
+# This recipe is originally from meta-openstack:
```
Can we reuse the one from meta-openembedded? https://layers.openembedded.org/layerindex/recipe/120595/

bhargavdas replied:
Same for this recipe

vj-kumar commented on meta-mtda/recipes-devtools/python/python3-zeroconf_0.36.13.bb:
```diff
@@ -0,0 +1,13 @@
+SUMMARY = "Pure Python Multicast DNS Service Discovery Library (Bonjour/Avahi compatible)"
```
https://layers.openembedded.org/layerindex/recipe/182205/

bhargavdas replied:
Again, same for this recipe also.

bhargavdas commented:
@vj-kumar all upstream dropped
https://github.com/siemens/mtda/pull/240/commits/91f3168d826200d6aa35630c1b7ec13646081034

initial sanity checks working as expected.

bhargavdas commented:
> > > > @bhargavdas Wondering if we can start with a proven mtda platform for the initial support. Beaglebone black might be a good candidate to start with. All you need is the poky repo.
> > > > Starting with such a platform will give us some idea on how much of things we can reuse, and how we can align our existing implementation to achieve that.
> > > 
> > > 
> > > I agree, but I don't have a beaglebone black to test. Though I can build an image using this layer and ask someone to test it.
> > 
> > 
> > I have one and I can lend it to you.
> 
> Sounds good :)

The layer is building with beaglebone-yocto image. Testing in progress.
Since this layer prepared sometime back, few services like `mtda-usb-functions` need to removed as it is merge now with mtda itself.

bhargavdas commented:
Initial checks on beaglebone black looks fine. 

service
```root@mtda:~# systemctl status mtda
● mtda.service - Multi-Tenant Device Access
     Loaded: loaded (/lib/systemd/system/mtda.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2022-05-05 04:25:20 UTC; 5s ago
   Main PID: 294 (python3)
      Tasks: 1 (limit: 1130)
     Memory: 15.0M
     CGroup: /system.slice/mtda.service
             └─294 python3 /usr/bin/mtda-cli -d -n

May 05 04:25:20 mtda systemd[1]: Started Multi-Tenant Device Access.
root@mtda:~# 
```
mtda-cli help
```
root@mtda:~# mtda-cli help
usage: mtda [options] <command> [<args>]

The most commonly used mtda commands are:
   command   Send a command (string) to the device
   console   Interact with the device console
   getenv    Get named variable from the environment
   monitor   Interact with the device monitor console (if any)
   target    Power control the device
   setenv    Set named variable to specified value in the environment
   storage   Interact with the shared storage device
   usb       Control USB devices attached to the device

root@mtda:~# 
```

Next tasks before dropping WIP tag
- Cleanup layer and reuse bits directly from this repo as @vj-kumar suggested
- Check and validate other features

bhargavdas commented:
Somehow `/sys/class/gpio` directory is not present in the system, hence mtda service fails :(

```
root@mtda:~# ls /sys/class/gpio
ls: /sys/class/gpio: No such file or directory
root@mtda:~# lsmod
Module                  Size  Used by
snd_soc_simple_card    16384  0
snd_soc_simple_card_utils    24576  1 snd_soc_simple_card
snd_soc_davinci_mcasp    32768  2
snd_soc_ti_udma        16384  1 snd_soc_davinci_mcasp
snd_soc_ti_edma        16384  1 snd_soc_davinci_mcasp
snd_soc_ti_sdma        16384  1 snd_soc_davinci_mcasp
snd_soc_hdmi_codec     16384  1
snd_soc_core          204800  7 snd_soc_davinci_mcasp,snd_soc_hdmi_codec,snd_soc_simple_card_utils,snd_soc_ti_sdma,snd_soc_ti_edma,snd_soc_ti_udma,snd_soc_simple_card
snd_pcm_dmaengine      16384  1 snd_soc_core
snd_pcm               106496  4 snd_soc_davinci_mcasp,snd_pcm_dmaengine,snd_soc_hdmi_codec,snd_soc_core
snd_timer              32768  1 snd_pcm
snd                    65536  4 snd_soc_hdmi_codec,snd_timer,snd_soc_core,snd_pcm
soundcore              16384  1 snd
sch_fq_codel           20480  2
fuse                  118784  1
configfs               40960  1
root@mtda:~# 
```

bhargavdas commented:
> Somehow `/sys/class/gpio` directory is not present in the system, hence mtda service fails :(
> 
> ```
> root@mtda:~# ls /sys/class/gpio
> ls: /sys/class/gpio: No such file or directory
> root@mtda:~# lsmod
> Module                  Size  Used by
> snd_soc_simple_card    16384  0
> snd_soc_simple_card_utils    24576  1 snd_soc_simple_card
> snd_soc_davinci_mcasp    32768  2
> snd_soc_ti_udma        16384  1 snd_soc_davinci_mcasp
> snd_soc_ti_edma        16384  1 snd_soc_davinci_mcasp
> snd_soc_ti_sdma        16384  1 snd_soc_davinci_mcasp
> snd_soc_hdmi_codec     16384  1
> snd_soc_core          204800  7 snd_soc_davinci_mcasp,snd_soc_hdmi_codec,snd_soc_simple_card_utils,snd_soc_ti_sdma,snd_soc_ti_edma,snd_soc_ti_udma,snd_soc_simple_card
> snd_pcm_dmaengine      16384  1 snd_soc_core
> snd_pcm               106496  4 snd_soc_davinci_mcasp,snd_pcm_dmaengine,snd_soc_hdmi_codec,snd_soc_core
> snd_timer              32768  1 snd_pcm
> snd                    65536  4 snd_soc_hdmi_codec,snd_timer,snd_soc_core,snd_pcm
> soundcore              16384  1 snd
> sch_fq_codel           20480  2
> fuse                  118784  1
> configfs               40960  1
> root@mtda:~# 
> ```

@chombourger @vj-kumar  , Do I need to add / load any modules or dts entries ? Need a starting point to start debugging.

vj-kumar commented:
> > Somehow `/sys/class/gpio` directory is not present in the system, hence mtda service fails :(
> > ```
> > root@mtda:~# ls /sys/class/gpio
> > ls: /sys/class/gpio: No such file or directory
> > root@mtda:~# lsmod
> > Module                  Size  Used by
> > snd_soc_simple_card    16384  0
> > snd_soc_simple_card_utils    24576  1 snd_soc_simple_card
> > snd_soc_davinci_mcasp    32768  2
> > snd_soc_ti_udma        16384  1 snd_soc_davinci_mcasp
> > snd_soc_ti_edma        16384  1 snd_soc_davinci_mcasp
> > snd_soc_ti_sdma        16384  1 snd_soc_davinci_mcasp
> > snd_soc_hdmi_codec     16384  1
> > snd_soc_core          204800  7 snd_soc_davinci_mcasp,snd_soc_hdmi_codec,snd_soc_simple_card_utils,snd_soc_ti_sdma,snd_soc_ti_edma,snd_soc_ti_udma,snd_soc_simple_card
> > snd_pcm_dmaengine      16384  1 snd_soc_core
> > snd_pcm               106496  4 snd_soc_davinci_mcasp,snd_pcm_dmaengine,snd_soc_hdmi_codec,snd_soc_core
> > snd_timer              32768  1 snd_pcm
> > snd                    65536  4 snd_soc_hdmi_codec,snd_timer,snd_soc_core,snd_pcm
> > soundcore              16384  1 snd
> > sch_fq_codel           20480  2
> > fuse                  118784  1
> > configfs               40960  1
> > root@mtda:~# 
> > ```
> 
> @chombourger @vj-kumar , Do I need to add / load any modules or dts entries ? Need a starting point to start debugging.

Is CONFIG_GPIO_SYSFS kernel config set? zcat /proc/config.gz|grep GPIO

bhargavdas commented:
> Is CONFIG_GPIO_SYSFS kernel config set? zcat /proc/config.gz|grep GPIO

After multiple rebuilds (with different kernels ) I found that `CONFIG_GPIO_SYSFS` is not present in the latest linux-yocto kernel `5.15.36` but it is present in `5.10.63`.

Initial gpio control tests `mtda-cli target on/off` seems work fine now with `5.10` kernel. 
Now looking into what made the config disappear.

bhargavdas commented:
Update: gpio controls validated by checking sysfs values of `gpio48/value` both using `mtda-cli` console and `mtda-cli target on/off`.

chombourger commented on meta-mtda/recipes-core/usb-otg/mtda-usb-functions.bb:
```diff
@@ -0,0 +1,24 @@
+DESCRIPTION        = "USB functions for MTDA"
```
no longer needing this service (mtda is now creating USBF devices by itself)

bhargavdas replied:
removed

chombourger commented on meta-mtda/recipes-core/usb-otg/mtda-usb-functions/mtda-usb-functions.service:
```diff
@@ -0,0 +1,11 @@
+[Unit]
```
no longer needing this service (mtda is now creating USBF devices by itself)

bhargavdas replied:
removed

chombourger commented on meta-mtda/recipes-core/usb-otg/mtda-usb-functions/usb-functions:
```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
```
no longer needing this service (mtda is now creating USBF devices by itself)

chombourger commented on meta-mtda/recipes-python/pduclient/pduclient.bb:
```diff
@@ -0,0 +1,19 @@
+
+
+FILESEXTRAPATHS:prepend := "${THISDIR}/${PN}:"
+SRC_URI = "file://pduclient"
```
pull from https://github.com/pdudaemon/pdudaemon

chombourger commented on meta-mtda/README.md:
```diff
@@ -0,0 +1,13 @@
+# MTDA Yocto layer
```
rename this folder to meta-yocto ?

bhargavdas replied:
renamed

chombourger commented:
- most files added by this PR need a header with SPDX and Copyright lines
- determine how to share resource files between meta-isar and meta-yocto
- add a ci job to make sure your work builds ok within the GH infrastructure

bhargavdas commented on meta-mtda/recipes-core/usb-otg/mtda-usb-functions/postinst:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
removed

bhargavdas commented on meta-mtda/recipes-python/pduclient/files/pduclient:
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/python3
```
Pull from upstream now

bhargavdas commented:
Currently there are lot of commits which needs to be squashed or dropped. Once everything is cleaned up and the review comments are addressed, I will remove the RFC tags. Will try to complete by end of this week.

bhargavdas commented:
Hi @chombourger , @vj-kumar , 
* I have rebased and squashed all the commits.
* Added license headers.
* Added a kas file for poky qemu arm.
* Added ci job based on the kas qemu arm.

Please check if any additional changes/improvements are needed.

**NOTE**: Currently the mtda recipes PV is git, and SRCREV set to AUTOREV. Whenever the development window for the next release closes we will have to rename the recipe as mtda_<version>.bb and fix the SRCREV to a commit.

chombourger commented on meta-yocto/README.md:
```diff
@@ -0,0 +1,13 @@
+    > bitbake mtda-image
+
+### Known issues
+1. While building new image after changing the value of `MTDA_MASS_STORAGE_FILE` in `local.conf` (if already a image was alredy built), errors related to `usb-ms-use-storage` can be observed. Currenty the work around is cleaning the `usb-ms-use-storage` package with `bitbake usb-ms-use-storage` and then building the image again.
```
`MTDA_MASS_STORAGE_FILE` was removed some time ago. update?

chombourger commented on meta-yocto/conf/layer.conf:
```diff
@@ -0,0 +1,24 @@
+DISTRO_FEATURES:append = " systemd "
+VIRTUAL-RUNTIME_init_manager = "systemd"
+
+PREFERRED_VERSION:mtda = "0.18"
```
we shouldn't need this. do we?

bhargavdas replied:
No, I was trying something out and forgot to remove it.

chombourger commented on meta-yocto/recipes-python/mtda/mtda_0.18.bb:
```diff
@@ -0,0 +1,49 @@
+
+SRCREV	 = "38b5c33d8c9f9c7d2ed668551f797821f0f99029"
+
+SRC_URI += "git://git@github.com/siemens/mtda.git;protocol=https;branch=master"
```
check how the mtda recipe for Isar is now fetching its source files

bhargavdas replied:
Ok, will adapt to that. Need to digest the process and then will move it here.

bhargavdas replied:
Recipe is fetching the sources similar to isar. 
But once https://github.com/siemens/mtda/pull/278 get merged the recipe has to be changed

bhargavdas replied:
Need some changes here.

chombourger commented:
CI fails on the following reuse errors:
```
The following files have no copyright and licensing information:
* kas/yocto/mtda-beaglebone-yocto.yml
* kas/yocto/mtda-qemu-yocto.yml
* meta-yocto/recipes-devtools/python/python3-daemon/0001-Workaround-for-issue-2-1.patch
```

bhargavdas commented on meta-yocto/recipes-python/mtda/mtda_git.bb:
```diff
@@ -0,0 +1,87 @@
+
+S = "${WORKDIR}/working-repo"
+
+do_gen_working_repo() {
```
@chombourger @vj-kumar Please review the SRI_URI changes.

bhargavdas commented:
Let's hold this PR for few more days, until python3-daemon recipe gets merged in meta-python

bhargavdas commented:
> Let's hold this PR for few more days, until python3-daemon recipe gets merged in meta-python

https://github.com/siemens/mtda/pull/240/commits/0f2c6814ff8a7bca60b991dd6ad315ec789ee3d5
Updated the  layer to support latest yocto release

bhargavdas commented:
@chombourger , @vj-kumar this PR is ready to merge now.

chombourger commented on meta-yocto/README.md:
```diff
@@ -0,0 +1,19 @@
+    > bitbake mtda-image
+
+### Building for Beaglebone black
+Imageis for BBB can be built Using `kas/yocto/mtda-beaglebone-yocto.yml`.
```
The image for the Beaglebone black may be built with ..

bhargavdas replied:
Done

chombourger commented on .github/workflows/main.yml:
```diff
@@ -101,3 +101,14 @@ jobs:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Build image
+        if: github.ref == 'refs/heads/master'
```
if you are wanting PR builds to check Yocto builds then this line should be removed

bhargavdas replied:
Oh understood now, then we can keep this I guess, I have already checked the builds. Checking each PR build might take quite a lot of time.

chombourger replied:
Any idea how long it would take? I am worried that Yocto support could easily get broken if not checked regularly
maybe we need to implement sstate caching like done in https://github.com/siemens/meta-coral ?

bhargavdas replied:
>1hr in a 16 core machine. But I believe the machines provide by github actions are quite low powered.
We can only build the mtda package instead of building the whole image. Let me check in a similar machine for the same , will get a rough idea.

bhargavdas replied:
I have added this stage (https://github.com/siemens/mtda/pull/240/commits/2f0a9f9f7770ffb9e2c40f4d9916a6919fcef220) to build only the mtda package, please assign a `check` tag to check how much time it takes.
I triggered a mtda package build yesterday in a fairly older machine the build took around 4hrs.

bhargavdas replied:
Looks like the space offered by the github ci machine are enough need find alternate methods :(

bhargavdas replied:
@chombourger @vj-kumar how should we handle the ci for yocto since github's runners run out of space. Shall we run locally once a week ? or hook a self hosted runner? Suggestions.

jluebbe replied:
@bhargavdas You should probably enable [rm_work](https://docs.yoctoproject.org/ref-manual/classes.html#rm-work), which will at least help with the space usage issue.

bhargavdas commented:
@chombourger , What should be the approach for this PR ? Since the runners provided by GH are not powerful enough shall I setup a build at my end locally and publish results weekly or monthly in some form here ?
Also looks like this PR needs rebase.
